### PR TITLE
unit test on nmq-tcp v5 & mqtt_parser & fix memleak in conf free

### PR DESCRIPTION
**Pull request by Hermann0222**

### CODE REVIEW

**Review by RanMaoyi — APPROVED** [review]

**Review by wanghaEMQ — APPROVED** [review]

**Review by RanMaoyi — CHANGES_REQUESTED** [review]
sizeof(conf_web_hook_rule *) 
means 
sizeof(struct conf_web_hook_rule *) = 4Byte / 8Byte

**Review by lee-emqx — APPROVED** [review]

File: src/sp/protocol/mqtt/mqtt_parser_test.c [hunk 1]
```diff
@@ -3,18 +3,49 @@
 #include <nuts.h>
 #include <stdio.h>
 
+static void test_pub_extra()
+{
+	pub_extra *extra = NULL;
+	extra = pub_extra_alloc(extra);
```
[the remark this answers is not in the document]
Comment (Hermann0222) [reply src/sp/protocol/mqtt/mqtt_parser_test.c:9]: you are right

File: src/sp/protocol/mqtt/mqtt_parser_test.c [hunk 2]
```diff
@@ -3,18 +3,49 @@
...
-	// test oversize src.
+	// test oversize & NULL src.
 	NUTS_FAIL(utf8_check((char *) src, 65537), ERR_INVAL);
+	NUTS_FAIL(utf8_check(NULL, 0), ERR_INVAL);
 	// test control characters.
 	src[0] = 0x04;
 	NUTS_FAIL(utf8_check((char *) src, strlen((char *) src) - 1),
 	    ERR_MALFORMED_UTF8);
+	// test malformed utf8
+	NUTS_FAIL(utf8_check((char *) src_0, strlen((char *) src) - 1),
```
[the remark this answers is not in the document]
Comment (Hermann0222) [reply src/sp/protocol/mqtt/mqtt_parser_test.c:41]: that is my mistake

File: src/supplemental/nanolib/conf_test.c [hunk 3]
```diff
@@ -20,7 +20,10 @@ get_test_conf(const char *conf_path)
 void
 test_get_size(void)
 {
-	char    *str_size1 = "3KB";
+	char    cwd[1024];
+	getcwd(cwd,strlen(cwd));
```
[the remark this answers is not in the document]
Comment (Hermann0222) [reply src/supplemental/nanolib/conf_test.c:24]: these are debug only, will be commented out soon

File: src/supplemental/nanolib/conf_ver2.c [hunk 4]
```diff
@@ -533,14 +533,19 @@ conf_web_hook_parse_rules_ver2(conf *config, cJSON *jso)
...
 
 	cJSON_ArrayForEach(jso_webhook_rule, jso_webhook_rules)
 	{
-		conf_web_hook_rule *hook_rule = NNI_ALLOC_STRUCT(hook_rule);
-		webhook_action_parse_ver2(jso_webhook_rule, hook_rule);
-		cvector_push_back(webhook->rules, hook_rule);
-		webhook->rule_count = cvector_size(webhook->rules);
+		cnt++;
+		webhook->rules = realloc(
+		    webhook->rules, cnt * sizeof(conf_web_hook_rule *));
```
[the remark this answers is not in the document]
Comment (Hermann0222) [reply src/supplemental/nanolib/conf_ver2.c:542]: this is how it was designed in the first place in conf.c

[the remark this answers is not in the document]
Comment (RanMaoyi) [reply src/supplemental/nanolib/conf_ver2.c:542]: Got it, my fault.

**Review by RanMaoyi — APPROVED** [review]
LGTM

File: src/supplemental/nanolib/conf_ver2.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (Hermann0222) [reply src/supplemental/nanolib/conf_ver2.c]: same as in conf.c gotta to fix that too

[the remark this answers is not in the document]
Comment (RanMaoyi) [reply src/supplemental/nanolib/conf_ver2.c]: > same as in conf.c gotta to fix that too

webhook->rules is a (void **) type. So it's current to alloc(sizeof(void *)). Do not need to fix that.

**Pull request merged**